crates/lind-multi-process/src/lib.rs

```diff
@@ -0,0 +1,1323 @@
+
+                    let invoke_res = child_start_func
+                        .call(&mut store, &values, &mut results);
+
+                    // print errors if any when running the child process
+                    if let Err(err) = invoke_res {
+                        let e = wasi_common::maybe_exit_on_error(err);
+                        eprintln!("Error: {:?}", e);
+                        return 0;
+                    }
+
+                    // get the exit code of the module
+                    let exit_code = results.get(0).expect("_start function does not have a return value");
+                    match exit_code {
+                        Val::I32(val) => {
+                            // exit the cage with the exit code
+                            lind_syscall_api(
+                                child_cageid,
+                                30 as u32,
+                                0,
+                                0,
+                                *val as u64,
+                                0,
+                                0,
+                                0,
+                                0,
+                                0,
+                            );
+                            // let _ = on_child_exit(*val);
+                        },
+                        _ => {
+                            eprintln!("unexpected _start function return type!");
+                        }
+                    }
+
+                    // the cage just exited, decrement the cage counter
+                    lind_manager.decrement();
+                }
+
+                return 0;
+            }).unwrap();
+
+            // wait until child has fully copied the memory
+            barrier.wait();
+
+            // mark the parent to rewind state
+            let _ = asyncify_start_rewind_func.call(&mut store, rewind_pointer as i32);
+
+            // set up rewind state and fork return value for parent
+            store.set_rewinding_state(RewindingReturn {
+                rewinding: true,
+                retval: child_cageid as i32,
+            });
+
+            // return InvokeAgain here would make parent re-invoke main
+            return Ok(OnCalledAction::InvokeAgain);
+        }));
+
+        // after returning from here, unwind process should start
+        return Ok(0);
+    }
+
+    // shared-memory version of fork syscall, used to create a new thread
+    // This is very similar to normal fork syscall, except the memory is not copied
+    // and the saved unwind context need to be carefully copied and managed since parent
+    // and child are operating two copies to unwind data in the same memory
+    // Function Argument:
+    // * stack_addr: child's base stack address
+    // * stack_size: child's stack size
+    // * child_tid: the address of the child's thread id. This should be set by wasmtime
+    pub fn fork_shared_call(&self, mut caller: &mut Caller<'_, T>,
```

Comment:
Is it possible to rename this function? If this is to create new threads and not "cages" it not really a fork right?